contrib/aerospike/aerospike-client.v5/aerospike.go

```diff
@@ -0,0 +1,450 @@
+// Copyright 2016 Datadog, Inc.
+
+// Package aerospike provides functions to trace the aerospike/aerospike-client-go package (https://github.com/aerospike/aerospike-client-go).
+//
+// `WrapClient` will wrap an aerospike `Client` and return a new struct with all
+// the same methods, so should be seamless for existing applications. It also
+// has an additional `WithContext` method which can be used to connect a span
+// to an existing trace.
+package aerospike // import "gopkg.in/DataDog/dd-trace-go.v1/contrib/aerospike/aerospike-client.v5"
+
+import (
+	"context"
+	"math"
+	"time"
+
+	as "github.com/aerospike/aerospike-client-go/v5"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+// WrapClient wraps an aerospike.Client so that all requests are traced using the
+// default tracer with the service name "aerospike".
+func WrapClient(client *as.Client, opts ...ClientOption) *Client {
+	cfg := new(clientConfig)
+	defaults(cfg)
+	for _, opt := range opts {
+		opt(cfg)
+	}
+	log.Debug("contrib/aerospike/aerospike-client.v5: Wrapping Client: %#v", cfg)
+	return &Client{
+		Client:  client,
+		cfg:     cfg,
+		context: context.Background(),
+	}
+}
+
+// A Client is used to trace requests to the aerospike server.
+type Client struct {
+	*as.Client
+	cfg     *clientConfig
+	context context.Context
+}
+
+// WithContext creates a copy of the Client with the given context.
+func (c *Client) WithContext(ctx context.Context) *Client {
+	// the existing aerospike client doesn't support context, but may in the
+	// future, so we do a runtime check to detect this
+	mc := c.Client
+	if wc, ok := (interface{})(c.Client).(interface {
+		WithContext(context.Context) *as.Client
+	}); ok {
+		mc = wc.WithContext(ctx)
+	}
+	return &Client{
+		Client:  mc,
+		cfg:     c.cfg,
+		context: ctx,
+	}
+}
+
+// startSpan starts a span from the context set with WithContext.
+func (c *Client) startSpan(resourceName string) ddtrace.Span {
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeMemcached),
+		tracer.ServiceName(c.cfg.serviceName),
+		tracer.ResourceName(resourceName),
+	}
+	if !math.IsNaN(c.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, c.cfg.analyticsRate))
+	}
+	span, _ := tracer.StartSpanFromContext(c.context, operationName, opts...)
+	return span
+}
+
+// wrapped methods:
+
+// Put writes record bin(s) to the server.
+// The policy specifies the transaction timeout, record expiration and how the transaction is
+// handled when the record already exists.
+// If the policy is nil, the default relevant policy will be used.
+func (c *Client) Put(policy *as.WritePolicy, key *as.Key, binMap as.BinMap) as.Error {
+	span := c.startSpan("Put")
+	err := c.Client.Put(policy, key, binMap)
+	span.Finish(tracer.WithError(err))
+	return err
```

Comment:
```suggestion
func (c *Client) Put(policy *as.WritePolicy, key *as.Key, binMap as.BinMap) (err as.Error) {
	span := c.startSpan("Put")
	defer func() { span.Finish(tracer.WithError(err)) }()
	return c.Client.Put(policy, key, binMap)
```
Would it not be wiser to do it like this? Just in case something happens in between and the span doesn't get closed.

Reply:
Well, that's how it's done for the memcache one. The only thing that can happen in between is a `panic()` but then you most likely have much more problems than an unclosed span. `defer` are executed during a panic so yeah it would be technically better.

Reply:
I see. In that case - don't you think we should do the same here? I know that historically this has been an issue. Better solve it now from the start.

Reply:
done